src/hp/derp/http/server.rs

```diff
@@ -1,50 +1,334 @@
-use std::pin::Pin;
-use std::task::{Context, Poll};
-
-use anyhow::Result;
-use futures::future::FutureExt;
-use futures::Future;
-use hyper::header::{HeaderValue, UPGRADE};
-use hyper::upgrade::Upgraded;
-use hyper::{Body, Request, Response, StatusCode};
-
-use tokio::net::tcp::{OwnedReadHalf, OwnedWriteHalf};
-use tracing::debug;
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use anyhow::{bail, ensure, Context as _, Result};
+use bytes::Bytes;
+use derive_more::Debug;
+use futures::future::{Future, FutureExt};
+use http::response::Builder as ResponseBuilder;
+use hyper::{
+    header::{HeaderValue, UPGRADE},
+    server::conn::Http,
+    upgrade::Upgraded,
+    Body, HeaderMap, Method, Request, Response, StatusCode,
+};
+use tokio::{
+    net::{TcpListener, TcpStream},
+    task::JoinHandle,
+};
+use tokio_rustls_acme::AcmeAcceptor;
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, debug_span, error, info, warn, Instrument};
 
 use super::HTTP_UPGRADE_PROTOCOL;
-use crate::hp::derp::{server::ClientConnHandler, types::PacketForwarder};
+use crate::hp::{
+    derp::{
+        http::client::Client as HttpClient, server::ClientConnHandler, server::MaybeTlsStream,
+        types::MeshKey, types::PacketForwarder, MaybeTlsStreamServer,
+    },
+    key::node::SecretKey,
+};
+
+type HyperError = Box<dyn std::error::Error + Send + Sync>;
+type HyperResult<T> = std::result::Result<T, HyperError>;
+
+fn downcast_upgrade(upgraded: Upgraded) -> Result<(MaybeTlsStream, Bytes)> {
+    match upgraded.downcast::<MaybeTlsStream>() {
+        Ok(parts) => Ok((parts.io, parts.read_buf)),
+        Err(_) => {
+            bail!("could not downcast the upgraded connection to MaybeTlsStream")
+        }
+    }
+}
 
 /// The server HTTP handler to do HTTP upgrades
-pub async fn derp_connection_handler<P>(
-    conn_handler: &ClientConnHandler<OwnedReadHalf, OwnedWriteHalf, P>,
+async fn derp_connection_handler<P>(
+    conn_handler: &ClientConnHandler<P>,
     upgraded: Upgraded,
 ) -> Result<()>
 where
     P: PacketForwarder,
 {
     debug!("derp_connection upgraded");
-    // get the underlying TcpStream
-    let parts = match upgraded.downcast::<tokio::net::TcpStream>() {
-        Ok(p) => p,
-        Err(_) => {
-            anyhow::bail!("could not downcast the upgraded connection to a tokio::net::TcpStream")
-        }
-    };
-    anyhow::ensure!(
-        parts.read_buf.is_empty(),
+    let (io, read_buf) = downcast_upgrade(upgraded)?;
+    ensure!(
+        read_buf.is_empty(),
         "can not deal with buffered data yet: {:?}",
-        parts.read_buf
+        read_buf
     );
 
-    // split into the reader and writer parts
-    let (reader, writer) = parts.io.into_split();
+    conn_handler.accept(io).await
+}
+
+#[derive(Debug)]
```

Comment:
odd place to put the derive ;)

Reply:
what you don't like it dangling around in the wind like that 😄 ?